packages/slate-editor/src/extensions/galleries/components/Gallery/Gallery.tsx

```diff
@@ -1,69 +1,274 @@
-import type { GalleryNode } from '@prezly/slate-types';
-import type { UploadcareImage } from '@prezly/uploadcare';
+import type { DragEndEvent, DragStartEvent, UniqueIdentifier } from '@dnd-kit/core';
+import {
+    closestCenter,
+    DndContext,
+    DragOverlay,
+    PointerSensor,
+    useSensor,
+    useSensors,
+} from '@dnd-kit/core';
+import { arrayMove, SortableContext } from '@dnd-kit/sortable';
+import { GalleryImageSize, type GalleryImage, type GalleryNode } from '@prezly/slate-types';
+import { awaitUploads, UploadcareImage } from '@prezly/uploadcare';
 import { isUploadcareImageSizeValid } from '@prezly/uploadcare';
-import classNames from 'classnames';
+import { noop } from '@technically/lodash';
 import type { HTMLAttributes } from 'react';
-import React from 'react';
+import React, { useCallback, useMemo, useState } from 'react';
+import { createPortal } from 'react-dom';
+import { useSlateStatic } from 'slate-react';
+
+import { UploadcareEditor } from '#modules/uploadcare';
 
 import { IMAGE_PADDING, IMAGE_SIZE } from './constants';
 import styles from './Gallery.module.scss';
 import { GalleryTile } from './GalleryTile';
+import type { Tile } from './lib';
 import { calculateLayout } from './lib';
+import { SortableGalleryTile } from './SortableGalleryTile';
 
 interface Props extends HTMLAttributes<HTMLDivElement> {
     className?: string;
-    images: UploadcareImage[];
+    images: GalleryImage[];
+    isInteractive: boolean;
+    maxViewportWidth?: number;
+    onImagesChange: (images: GalleryImage[]) => void;
     padding: GalleryNode['padding'];
     size: GalleryNode['thumbnail_size'];
+    uuid: string;
     width: number;
-    maxViewportWidth?: number;
 }
 
 export function Gallery({
     className,
-    images,
+    images: originalImages,
+    isInteractive,
     maxViewportWidth = 800,
+    onImagesChange,
     padding,
     size,
+    uuid,
     width,
     ...attributes
 }: Props) {
+    const editor = useSlateStatic();
+    const [activeId, setActiveId] = useState<UniqueIdentifier | null>(null);
+    const images = useMemo(
+        () =>
+            originalImages.map((image, index) => ({ ...image, id: `${image.file.uuid}-${index}` })),
```

Comment:
Since we rely on Uploadcare's UUID for keys and drag & drop ids, it stops working correctly as soon as you upload the same image twice, creating a UUID conflict.

I've spoked about this with Al and we believe it is okay to remove duplicate images from the gallery, but I am not 100 % sure to pull the trigger. We could write a normalization for this.

So, for now, I've implemented this extra logic that generates a unique `id` for each image in the gallery. This way, we can still support duplicate images, but it makes the code a bit more complex.

A small downside of this approach is the fact that I'm using index to generate the id, so some re-rendering of gallery tiles will be unavoidable, since the index will change for each image in the array that has to change a position.

Reply:
> so ideally I still think we should remove duplicate images so we can rely on just the Uploadcare's UUID

I don't think this is a good idea.

There can be different images with the same `uuid`. Imagine uploading a large photo with 3 people, but then cropping it to a different person every time. This way you'll have three different photos with exactly the same `uuid`.

--- 

I believe the best way to fix it would be to stop relying on `image.file.uuid` as an identifier and instead add a new  autogenerated unique`image.uuid` property. Not for now, of course.